module/zfs/dmu_objset.c

```diff
@@ -942,13 +984,25 @@ dmu_objset_create_sync(void *arg, dmu_tx_t *tx)
 	VERIFY0(dsl_dir_hold(dp, doca->doca_name, FTAG, &pdd, &tail));
 
 	obj = dsl_dataset_create_sync(pdd, tail, NULL, doca->doca_flags,
-	    doca->doca_cred, tx);
+	    doca->doca_cred, doca->doca_dcp, tx);
 
 	VERIFY0(dsl_dataset_hold_obj(pdd->dd_pool, obj, FTAG, &ds));
 	bp = dsl_dataset_get_blkptr(ds);
 	os = dmu_objset_create_impl(pdd->dd_pool->dp_spa,
 	    ds, bp, doca->doca_type, tx);
 
+	/*
+	 * At this time, the objset is not owned and so it does not have its
+	 * key mapping in the keystore. This mapping might be needed to create
+	 * encrypted objects in the dataset. We create it here manually, but we
+	 * cannot destroy it until after syncing is complete. Therefore any
+	 * function that calls this must clean up the key mapping after syncing.
```

Comment:
Not exactly, although this is probably the most fragile part of the implementation so far. I couldn't come up with a way to make it better without significant code restructuring or introducing more fragility. Perhaps you can think of something better?

The key mappings are stored in the spa keystore whenever the dataset is owned, indexed by their dataset object number. These are used by the zio layer to lookup the master encryption keys. Normally this isn't a problem because anybody who wants to read or write encrypted data to a dataset must own it first, which loads the keys. However, in `dmu_objset_create()` there is no dataset to own yet. It does not start to exist until midway through `dmu_objset_create_sync()`, so we need to wait until that time to create the mapping. This mapping cannot be dropped until all the data from the sync function has been written out, which by definition won't happen until after the sync function has returned. Hence, the reason for the comment.